### PR TITLE
Add new character type

### LINKED ISSUE #375: Add a flag to the Index Block that indicates whether a data block encodes text
## Summary 
Currently in GSD, log data seems to be encoded in such a way that it is not possible to determine whether it should be decoded as numerical or textual information. This can be seen in a MWE provided below, in which a GSD file is written with Hoomd, but upon reading with GSD's own built-in functions, one of the logged fields seems (at first glance) to be decoded incorrectly. This behavior is actually described in the docs, but it is difficult to find that description, and the behavior can feel very unexpected to newcomers. I suggest that we add a flag to the index block to indicate whether a block contains numerical or textual information, and then modify the behavior of GSD's built-in reading functions to behave accordingly.

## Description
Here is a MWE of a simple simulation that writes trajectories and a message "hello" to an output file called `test.gsd`:
```python
import hoomd
import gsd.hoomd

# Initialize Frame
frame = gsd.hoomd.Frame()
frame.particles.types = ["A"]

positions = list(itertools.product(numpy.linspace(-2, 2, 2, endpoint=False), repeat=3))
frame.particles.N = len(positions)
frame.particles.position = positions
frame.particles.typeid = [0] * frame.particles.N
frame.configuration.box = [8, 8, 8, 0, 0, 0]
frame.particles.mass = [1] * frame.particles.N
frame.particles.moment_inertia = [1, 1, 1] * frame.particles.N
frame.particles.orientation = [(1, 0, 0, 0)] * frame.particles.N

# Initialize Simulation
simulation = hoomd.Simulation(device=hoomd.device.CPU(), seed=1)
simulation.create_state_from_snapshot(frame)

# Create integrator
integrator = hoomd.md.Integrator(dt=0.001)
simulation.operations.integrator = integrator

method = hoomd.md.methods.Langevin(filter=hoomd.filter.All(), kT=1.0)

integrator.methods.append(method)

# Add logger and writer
logger = hoomd.logging.Logger()
logger[("message")] = (lambda: "hello", hoomd.logging.LoggerCategories.string)   #   <-- string data gets logged here
gsd_writer = hoomd.write.GSD(trigger=hoomd.trigger.Periodic(1),
    filename="test.gsd",
    logger=logger,
    filter=hoomd.filter.All()
)
simulation.operations.writers.append(gsd_writer)

# Run the simulation and flush the writer
simulation.run(3)
simulation.operations.writers[0].flush()
```
When `test.gsd` is opened in Ovito, the `message` field is displayed properly:

image

But when `test.gsd` is read with GSD's built-in functions, `message` instead contains arrays of integers:

```
print(gsd.hoomd.read_log("test.gsd"))
>>> {'configuration/step': array([1, 2, 3], dtype=uint64), 'log/message': array([[104, 101, 108, 108, 111,   0],
>>>       [104, 101, 108, 108, 111,   0],
>>>       [104, 101, 108, 108, 111,   0]], dtype=int8)}

with gsd.hoomd.open("test.gsd") as f:
    print(f[0].log["message"])
>>> [104 101 108 108 111   0]
```

These integers are the decimal form of [Unicode names](https://en.wikipedia.org/wiki/List_of_Unicode_characters#Basic_Latin) for the letters 'h', 'e', 'l', 'l', 'o', and `null`. Indeed, you can decode them this way:

```
messages = gsd.hoomd.read_log("test.gsd")["log/message"]
messages = [str(m, encoding="UTF-8").rstrip("\x00") for m in messages]
print(messages)
>>> ['hello', 'hello', 'hello']
```

After discussing this offline with @tcmoore3 , I learned that this behavior is not only expected, but also [documented](https://gsd.readthedocs.io/en/v3.3.0/fl-examples.html#store-string-chunks) - still, this behavior seems strange to me. I understand why data is stored the way it is in GSD, as blobs are much more efficient for read/write tasks than plain text, but the Python API is intended to be approachable, allowing users to, for example, create a pandas dataframe from a GSD log in [one line](https://gsd.readthedocs.io/en/v3.3.0/python-module-gsd.hoomd.html#gsd.hoomd.read_log). If we want to make it easy for users to interact with GSD files through Python, I think we should make it so they don't have to convert decimal unicode to text on their own. Just as an example, the oneliner I linked above will break if the log contains varying textual information, because pandas requires uniform dimensions in each column.

## Proposed solution

I am not familiar enough with GSD's [specification](https://gsd.readthedocs.io/en/v3.3.0/specification.html) and [C](https://gsd.readthedocs.io/en/v3.3.0/c-api.html) and [Python](https://gsd.readthedocs.io/en/v3.3.0/python-api.html) APIs to give detailed suggestions for a solution, but I do have some ideas. The file layer specification [indicates](https://gsd.readthedocs.io/en/v3.3.0/file-layer.html#index-block) that the Index Block contains an unused(?) line item called `flags`, which is "reserved for future use". I suggest that we consider adding a flag to indicate that a data block contains textual information.

If we implemented this flag, we would also need to modify the behavior of `gsd.hoomd.open()` and `gsd.hoomd.read_log()` to check for the flag and then automatically decode the decimal unicode. We would also need to add some logic in Hoomd handle the new behavior. Again, I am not familiar enough with Hoomd's internals to suggest a full solution, but this could involve modifying the behavior of `hoomd.write.GSD` to set the flag when a logger passes it string values.

### PR DESCRIPTION
**Pull request by josephburkhart**
## Description

These changes make the Python API more approachable by adding support for reading and writing strings to GSD data chunks. My changes add a new character data type on the C side, and modify the python api to automatically decode strings for the user. The new changes are accompanied by a new test, which fully covers the expected use cases.

## Motivation and Context

Currently, strings are implicitly encoded as uint8 data, but it is up to the user to figure out how to decode that data upon reading. The Python API is intended to be approachable, allowing users to, for example, create a pandas dataframe from a GSD log in [one line](https://gsd.readthedocs.io/en/v3.3.0/python-module-gsd.hoomd.html#gsd.hoomd.read_log). However, this oneliner will break if the log contains varying textual information, because pandas requires uniform dimensions in each column. These changes make it so users don't have to convert decimal unicode to text on their own, allowing the one-line example to work.

This PR resolves #375.

## How Has This Been Tested?

I have modified `test_fl.py`, adding a new test specifically for the string dtype (`test_string_dtype`). The new test covers both `fl` and `pygsd`.

- [x] Please build the sphinx documentation and check that any changes to documentation display properly

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/gsd/blob/trunk-patch/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**GSD Contributor Agreement**](https://github.com/glotzerlab/gsd/blob/trunk-patch/ContributorAgreement.md).
- [x] My name is on the list of contributors (`doc/credits.rst`) in the pull request source branch.
- [x] I have added a change log entry to ``CHANGELOG.rst``.

### CODE REVIEW

**Comment by josephburkhart** [comment]
pre-commit.ci autofix

File: gsd/fl.pyx [hunk 1]
```diff
@@ -22,6 +22,7 @@ from libc.stdint cimport uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t,\
 from libc.errno cimport errno
 cimport gsd.libgsd as libgsd
 cimport numpy
+cimport cython
```
Comment (joaander) [inline gsd/fl.pyx]:
```suggestion
```
Unused.

File: gsd/fl.pyx [hunk 2]
```diff
@@ -145,6 +146,21 @@ cdef void * __get_ptr_float64(data):
...
+    # data_array_char = data
+    # cdef char *data_array_char
+    # data_array_char = cython.address(data.data[0])
+    if (data.size == 0):
+        return NULL
+    else:
+        # return <void*>cython.address(data_array_char.data[0])
+        # return <void*>cython.address(data_array_char[0])
+        return <void*>address
+
```
Comment (joaander) [inline gsd/fl.pyx]:
```suggestion
```
Unused.

File: gsd/fl.pyx [hunk 3]
```diff
@@ -557,59 +573,84 @@ cdef class GSDFile:
...
+                data_ptr = __get_ptr_float32(data_array)
+            elif data_array.dtype == numpy.float64:
+                gsd_type = libgsd.GSD_TYPE_DOUBLE
+                data_ptr = __get_ptr_float64(data_array)
+            # elif data_array.dtype.type is numpy.str_:
+            #     if N != 0:
+            #         M = int(data_array.dtype.itemsize/4)
+            #     print(f"Wrote: {N=}, {M=}, {data_array=}")
+            #     gsd_type = libgsd.GSD_TYPE_CHARACTER
+            #     data_ptr = __get_ptr_char(data_array)
```
Comment (joaander) [inline gsd/fl.pyx]:
```suggestion
```

File: gsd/fl.pyx [hunk 4]
```diff
@@ -826,6 +872,12 @@ cdef class GSDFile:
             __raise_on_error(retval, self.name)
 
         if index_entry.M == 1:
+            if gsd_type == libgsd.GSD_TYPE_CHARACTER:
+                data_array = data_array.flatten()
+                bytes_array = data_array.view(dtype=numpy.dtype((bytes, data_array.shape[0])))
+                print(bytes_array[0].decode("UTF-8"))
```
Comment (joaander) [inline gsd/fl.pyx]:
```suggestion
```

File: gsd/gsd.c [hunk 5]
```diff
@@ -1607,6 +1613,24 @@ inline static int gsd_initialize_handle(struct gsd_handle* handle)
     handle->maximum_write_buffer_size = GSD_DEFAULT_MAXIMUM_WRITE_BUFFER_SIZE;
     handle->index_entries_to_buffer = GSD_DEFAULT_INDEX_ENTRIES_TO_BUFFER;
 
+    // When opening a file in a writeable mode, if the file's major version
+    // is identical to the current major version, silently upgrade the minor version
+    if ((handle->open_flags == GSD_OPEN_READWRITE || handle->open_flags == GSD_OPEN_APPEND)
+        && (handle->header.gsd_version
+            != gsd_make_version(GSD_CURRENT_FILE_VERSION_MAJOR, GSD_CURRENT_FILE_VERSION_MINOR))
+        && (handle->header.gsd_version >> (sizeof(uint32_t) * 4) == GSD_CURRENT_FILE_VERSION_MAJOR))
+        {
```
Comment (joaander) [inline gsd/gsd.c:1622]:
```suggestion
    if ((handle->open_flags == GSD_OPEN_READWRITE || handle->open_flags == GSD_OPEN_APPEND)
        && (handle->header.gsd_version
            != gsd_make_version(GSD_CURRENT_FILE_VERSION_MAJOR, GSD_CURRENT_FILE_VERSION_MINOR))
        && (handle->header.gsd_version >> (sizeof(uint32_t) * 4) == GSD_CURRENT_FILE_VERSION_MAJOR)
        && (handle->header.gsd_version & 0xFFFF < GSD_CURRENT_FILE_VERSION_MINOR))
        {
```
There is no need to rewrite the header when it is already 2.1.

File: gsd/gsd.c [hunk 6]
```diff
@@ -2554,8 +2582,9 @@ int gsd_upgrade(struct gsd_handle* handle)
                 }
             }
 
-        // label the file as a v2.0 file
-        handle->header.gsd_version = gsd_make_version(GSD_CURRENT_FILE_VERSION, 0);
+        // label the file as a v2.1 file
```
Comment (joaander) [inline gsd/gsd.c]:
```suggestion
        // GSD always writes files matching the current major and minor version.
```

**Review by joaander — CHANGES_REQUESTED** [review]
We need to remove stray `print()` calls, commented code, and unused imports.

**Comment by josephburkhart** [comment]
Great catches - apologies for the omissions!

File: gsd/gsd.c (hunk 5 above)
Comment (joaander) [reply gsd/gsd.c:1622]: I reverted this redundant suggestion. During the review I failed to notice that we already checked `(handle->header.gsd_version != gsd_make_version(GSD_CURRENT_FILE_VERSION_MAJOR, GSD_CURRENT_FILE_VERSION_MINOR)`.

**Review by joaander — APPROVED** [review]
Thanks!

**Comment by joaander** [comment]
I tested the original motivation for this feature.

In the `test_hoomd.py::read_log` test, pandas is indeed able to create a `DataFrame` with the strings:
```python
    import pandas
    df = pandas.DataFrame(logged_data_dict)
    print()
    print(df)
```

```
   configuration/step  log/value/potential_energy  log/value/pressure log/category
0                   0                          10                  -3            A
1                   1                          10                   5          BBB
```

**Pull request merged**